Rework the message handling

whot opened the pull request:
This is on top of #150 

This is a large rework of how we talk to the device, with the main goal of separating out all the device-specific messages into a separate module and their own classes. This should make a few things easier long-term, specifically testing and debugging.

The pen data handling is still on the old approach, this only switches "normal" messages over. Registration works with the new message system too but it's a bit clunky.

Tested with the Intuos Pro and the Spark, both still work here.

whot committed:
wacom: add another nordic function that just takes and returns data

This function doesn't process the data beyond the logging required. It simply
sends the request and returns the reply, whatever that is.

Currently unused, this is prep work for later.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: start a new attempt at handling the protocol

This is only the framework for a new protocol handler that should simplify
adding, extending and debugging the actual protocol messages.

The summary is:
- each protocol interaction is a subclass of Msg() that handles the data sent
  and received from the device.
- the list of messages for each device is stored in a dictionary the caller
  initializes by providing the protocol version
- those messages are addressable via the Interactions enum through the
  Protocol class
- the messages can be run via execute(), taking and setting the obvious
  properties on the object
- actual communication is done via a single callback that takes/returns a
  NordicData object for the request/reply.

So after the basic setup, the caller can do this:
   p = Protocol(ProtocolVersion.INTUOS_PRO)
   name = p.execute(Interactions.GET_NAME).name
   p.execute(Interactions.SET_TIME, time.time())

This completely separates the protocol from the implementation where we want
the result of that exchange, ideally paving the way for better testing and
easier debugging of what's happening on the wire. Well, ether.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: allow for "no specific requested opcode" where we just return the data

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: drop our NordicData, use the one from the protocol

Same thing anyway, this is just namespace handling

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: use the new protocol implemenation for all "normal" messages

This is an in-situ replacement of the old functionality vs the new one,
resulting in a bit of duplication. Now that the Protocol takes care of the
device-specifics most of our functions look identical. This will be fixed in a
follow-up commit.

Missing from the conversion here is the initial register handling (too
convoluted right now) and the parsing of the pen data.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: drop a bunch of now-duplicate functions

This is all handled by the matching Msg now.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: pass the timestamp in seconds down parse_pen_data

We don't carea about the struct time here

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: add a timeout argument to the new protocol handlers

The default timeout is too short for registering a button, let's control this
from the message.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: switch registering the device to the new protocol handling

This is still a bit awkward because the communication of the spark especially
is so strange. It sends a (rejected) uuid, then a single command to register
that UUID. The Slate and the IntuosPro are more sensible, they seem to have a
special message to register a UUID - that one either fails or succeeds.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
wacom: drop the now-obsolete send_nordic_command_sync

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
Drop the semi-duplicate separate Protocol class

We can hook into the ProtocolVersion here, so let's de-duplicate this.
Should be backwards-compatible for config files but new config files will have
the uppercase name written now.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
base: add a flag to control the config directory

For testing it's a lot easier to just provide a /tmp/ directory than having to
clear out the normal one that may contain useful drawings for debugging.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: drop the UNKNOWN_B1 class

0xb1 is the SetMode command, the rest is just the argument for which mode we
want

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: 0xec is simply the GATT selection for file transfers

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
protocol: split GET_DIMENSIONS up into two messages

It's the same message with a different argument, let's try not to do
gymnastics in the Msg code just so we can return a tuple.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

bentiss commented:
A rough view at the code is OK from my point of view.

However, I have an error when fetching new data:

```
15:11:48 DEBUG: tuhi.dbus: Listening started on glamorous sheep for :1.123
15:11:48 DEBUG: tuhi.ble: /org/bluez/hci0: Discovery started (timeout 0)
15:11:56 DEBUG: tuhi: E0:61:C6:BF:14:4E: UUID f09e6fecf1ba protocol: slate
15:11:56 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Connecting
15:11:56 DEBUG: tuhi: E0:61:C6:BF:14:4E: UUID f09e6fecf1ba protocol: slate
15:11:56 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Connecting
15:11:56 DEBUG: tuhi.ble.E0:61:C6:BF:14:4E: Already connecting
15:11:56 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Connection established
15:11:57 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX SYSEVENT <-- b'f1024a00'
15:11:57 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX SYSEVENT <-- b'f1024a00'
15:11:57 DEBUG: tuhi: E0:61:C6:BF:14:4E: connected for DeviceMode.LISTEN
15:11:57 DEBUG: tuhi.wacom: E0:61:C6:BF:14:4E: starting
15:11:57 DEBUG: tuhi.fw: command: check connection
15:11:57 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> e6 / 06 / f0 9e 6f ec f1 ba
15:11:57 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b3 01 00
15:11:57 DEBUG: tuhi.fw: command: set time
15:11:57 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> b6 / 06 / 19 08 12 13 11 57
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b3 01 00
15:11:58 DEBUG: tuhi.fw: command: read battery info
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> b9 / 01 / 00
15:11:58 DEBUG: tuhi.dbus: Listening stopped on glamorous sheep for :1.123
15:11:58 DEBUG: tuhi.ble: /org/bluez/hci0: Discovery stopped
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- ba 02 4a 00
15:11:58 INFO: tuhi.wacom: device battery: 74% (discharging)
15:11:58 DEBUG: tuhi.fw: command: get dimensions
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> ea / 02 / 03 00
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- eb 06 03 00 60 54 00 00
15:11:58 DEBUG: tuhi.fw: command: get dimensions
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> ea / 02 / 04 00
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- eb 06 04 00 d0 39 00 00
15:11:58 INFO: tuhi.wacom: dimensions: 21600x14800
15:11:58 DEBUG: tuhi.fw: command: get firmware
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> b7 / 01 / 00
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b8 09 00 0b 01 06 00 08 00 05 0a
15:11:58 DEBUG: tuhi.fw: command: get firmware
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> b7 / 01 / 01
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b8 09 01 0c 01 06 00 09 02 03 0a
15:11:58 INFO: tuhi.wacom: firmware is b160805a-c160923a
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> ec / 06 / 06 00 00 00 00 00
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b3 01 00
15:11:58 DEBUG: tuhi.fw: command: start/stop live
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> b1 / 01 / 01
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b3 01 00
15:11:58 DEBUG: tuhi.fw: command: start/stop live
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> b1 / 01 / 01
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- b3 01 00
15:11:58 DEBUG: tuhi.fw: command: check for data
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> c1 / 01 / 00
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- c2 02 03 00
15:11:58 DEBUG: tuhi.wacom: Drawings available: 3
15:11:58 DEBUG: tuhi.fw: command: fetch data
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: TX NORDIC --> cc / 01 / 00
15:11:58 DEBUG: tuhi.fw: E0:61:C6:BF:14:4E: RX NORDIC <-- cf 0a 9b 07 00 00 19 08 08 08 57 58
15:11:58 INFO: tuhi.ble.E0:61:C6:BF:14:4E: Disconnecting
Exception in thread Thread-4:
Traceback (most recent call last):
  File "/usr/lib64/python3.7/threading.py", line 917, in _bootstrap_inner
    self.run()
  File "/usr/lib64/python3.7/threading.py", line 865, in run
    self._target(*self._args, **self._kwargs)
  File "/home/btissoir/Src/tuhi/tuhi/wacom.py", line 1244, in _run
    self._wacom_protocol.retrieve_data()
  File "/home/btissoir/Src/tuhi/tuhi/wacom.py", line 1020, in retrieve_data
    if self.read_offline_data() == 0:
  File "/home/btissoir/Src/tuhi/tuhi/wacom.py", line 892, in read_offline_data
    count, timestamp = self.get_stroke_data()
  File "/home/btissoir/Src/tuhi/tuhi/wacom.py", line 755, in get_stroke_data
    msg = self.p.execute(Interactions.GET_STROKES)
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 235, in execute
    return self.get(key, *args, **kwargs).execute()
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 456, in execute
    self._handle_reply(self.reply)
  File "/home/btissoir/Src/tuhi/tuhi/protocol.py", line 905, in _handle_reply
    t = time.strptime(str_timestamp, '%y%m%d%H%M%S')
  File "/usr/lib64/python3.7/_strptime.py", line 571, in _strptime_time
    tt = _strptime(data_string, format)[0]
  File "/usr/lib64/python3.7/_strptime.py", line 359, in _strptime
    (data_string, format))
ValueError: time data '9b070000190808085758' does not match format '%y%m%d%H%M%S'
```

whot commented:
yeah, sorry, was a typo here, diff to fix this is squashed in now
```diff
diff --git a/tuhi/protocol.py b/tuhi/protocol.py
index 160f293..a70a704 100644
--- a/tuhi/protocol.py
+++ b/tuhi/protocol.py
@@ -875,7 +875,7 @@ class MsgGetStrokesSlate(Msg):
             raise UnexpectedReply(reply)
 
         self.count = int.from_bytes(reply[0:4], byteorder='little')
-        str_timestamp = ''.join([f'{d:02x}' for d in reply])
+        str_timestamp = ''.join([f'{d:02x}' for d in reply[4:]])
         t = time.strptime(str_timestamp, '%y%m%d%H%M%S')
         self.timestamp = calendar.timegm(t)
```

whot commented:
and rebased onto master now too

whot committed:
wacom: time_to_bytes is obsolete now

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot committed:
test: add unit-testing for the various protocol messages

This only tests for the success cases, but at least we know that path is
executed.

Signed-off-by: Peter Hutterer <peter.hutterer@who-t.net>

whot commented:
Ok, this is it now, I think. With unit-tests and thus a bunch of fixes squashed into the earlier patches.

bentiss reviewed (approved):
AFAICT, the unit tests are not run in circle CI.
Is it a container limitation? I am under the impression that you are emulating the Nordic callbacks, so you should be able to run them in CircleCI...

Otherwise, looks good.

whot commented:
Fixed, `pytest-3` was missing on the CI and it skipped the tests automatically.
